kernel/aster-nix/src/net/socket/ip/datagram/mod.rs

```diff
@@ -280,12 +280,14 @@ impl Socket for DatagramSocket {
     fn recvfrom(&self, buf: &mut [u8], flags: SendRecvFlags) -> Result<(usize, SocketAddr)> {
         debug_assert!(flags.is_all_supported());
 
-        poll_ifaces();
-        if self.is_nonblocking() {
+        let recv_bytes = if self.is_nonblocking() {
             self.try_recvfrom(buf, flags)
         } else {
             self.wait_events(IoEvents::IN, || self.try_recvfrom(buf, flags))
```

Comment:
Should we do another poll if `self.try_recvfrom(buf, flags)` returns `Ok(0)`?

Reply:
Sorry, I don't understand this comment. Do you mean that some `poll_interfaces()` are missing or redundant?

Reply:
I'm just curious whether we should do another `poll_ifaces` in `wait_events`, like
```rust
self.wait_events(IoEvents::IN, || {
        let res = self.try_recvfrom(buf, flags);
        poll_ifaces();
        res
    }
)
```

If `try_recvfrom` returns `Ok(0)`, `wait_events` will go to sleep. Should we do a `poll_iface` before the thread sleeps?

Referred to your provided comment,
> You need to poll after "something" has happened:
> - You called any method from sockets or the interface

Reply:
What do you think of the updated implementation?

Reply:
The new version looks good to me!

Reply:
Sorry, but I just realized that it might not address your concern.

Some clarification:
 - `wait_event` will go to sleep if and only if `try_recvfrom` returns `Err(EAGAIN)`.
 - **Case 1:** `try_recvfrom` will return `Err(EAGAIN)` if smoltcp's `recv` method returns `Ok(0)`.
 - **Case 2:** `try_recvfrom` will return `Ok(0)` if smoltcp's `recv` method returns `Err(Finished)`.
https://github.com/asterinas/asterinas/blob/81cca422058bff9771927293f9741548a8d62c67/kernel/aster-nix/src/net/socket/ip/stream/connected.rs#L58-L65

It is unclear whether your meaning is for Case 1 or Case 2, and the new version still does not cover both cases.

If we want to cover both cases, I cannot think of a better way than to add a new set of methods called `try_*_and_poll`, do you have better ideas?

Reply:
OK, the new implementation covers both cases. Please check it out. Thanks!

Reply:
> `wait_event` will go to sleep if and only if `try_recvfrom` returns Err(EAGAIN).

Sorry, originally I thought `wait_event` will go to sleep if `try_recvfrom` returns `Ok(0)`, this is a mistake. 

Well, considers the two cases you mentioned, I think the better place to do `poll_ifaces` is in `Connected::try_recvfrom`, just before the code you mentioned. 

```rust
pub fn try_recvfrom(&self, buf: &mut [u8], flags: SendRecvFlags) -> Result<usize> {
    let result = self
        .bound_socket
        .raw_with(|socket: &mut RawTcpSocket| socket.recv_slice(buf));
    
    // do poll ifaces here
    poll_ifaces();

    match result {
        Ok(0) => return_errno_with_message!(Errno::EAGAIN, "the receive buffer is empty"),
        Ok(recv_bytes) => Ok(recv_bytes),
        Err(RecvError::Finished) => Ok(0),
        Err(RecvError::InvalidState) => {
            return_errno_with_message!(Errno::ECONNRESET, "the connection is reset")
        }
    }
}
```

The problem may be that if we do `poll_ifaces` here, the read lock of `state` is `StreamSocket` is held when doing `poll_ifaces`. But it may not be a problem, since the `on_events` of `StreamSocket` will only acquire the read lock of `state` if the socket is connected. If `poll_ifaces` will use some `Mutex` internally, then we can replace the lock of `state` from `RwLock` to `RwMutex`.

Reply:
> The problem may be that if we do `poll_ifaces` here, the read lock of `StreamSocket.state` is held when doing `poll_ifaces`.

Well, I think calling `poll_ifaces` while holding the socket lock is not good practice. My concern is that `poll_ifaces` will poll all interfaces and all sockets, and if there are a number of interfaces and sockets, this can take a long time. Ideally, I think `poll_ifaces` should only poll the associated socket here, but such an API is not currently available.

> But it may not be a problem, since the `StreamSocket::on_events`(witch will be called in `poll_ifaces`) will only acquire the read lock of `StreamSocket.state` if the socket is connected.

Actually, I think it's pretty wired. If we replace `RwLock` with `SpinLock`, it causes deadlocks. Although there are no deadlocks with `RwLock`, there are some risks and some maintenance overhead because the lock usage is harder to understand, but I think it can be avoided.

Reply:
Thanks for you explanation! The current change is also acceptable for me.